contracts/vanchor/lib.rs

```diff
@@ -118,6 +212,259 @@ mod vanchor {
+            &mut self,
+            src_chain_id: u64,
+            root: [u8; 32],
+            latest_leaf_index: u32,
+            target: [u8; 32],
+        ) -> Result<()> {
+            let edge = Edge {
+                chain_id: src_chain_id,
+                root,
+                latest_leaf_index,
+                target,
+            };
+
+            self.linkable_tree.update_edge(edge);
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn transact_deposit(
+            &mut self,
+            proof_data: ProofData,
+            ext_data: ExtData,
+            recv_token_addr: AccountId,
+            recv_token_amt: u128,
+        ) -> Result<()> {
+            if self.tokenwrapper_addr != recv_token_addr {
+                return Err(Error::Unauthorized);
+            }
+
+            self.validate_proof(proof_data.clone(), ext_data.clone());
+
+            let ext_data_fee: u128 = ext_data.fee.clone();
+            let ext_amt: i128 = ext_data.ext_amount.parse().expect("Invalid ext_amount");
+            let abs_ext_amt = ext_amt.unsigned_abs();
+
+            let is_withdraw = ext_amt.is_negative();
+
+            if is_withdraw {
+                return Err(Error::InvalidExecutionEntry);
+            } else {
+                if abs_ext_amt > self.max_deposit_amt {
+                    return Err(Error::InvalidDepositAmount);
+                };
+                if abs_ext_amt != recv_token_amt {
+                    return Err(Error::InsufficientFunds);
+                };
+            }
+
+            let fee_exists = ext_data_fee != 0;
+
+            if fee_exists {
+                if self
+                    .env()
+                    .transfer(ext_data.relayer.clone(), ext_data_fee)
+                    .is_err()
+                {
+                    panic!("{}", ERROR_MSG);
+                }
+            }
+
+            self.execute_insertions(proof_data.clone());
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn transact_withdraw(
+            &mut self,
+            proof_data: ProofData,
+            ext_data: ExtData,
+        ) -> Result<()> {
+            self.validate_proof(proof_data.clone(), ext_data.clone());
+
+            let ext_data_fee: u128 = ext_data.fee.clone();
+            let ext_amt: i128 = ext_data.ext_amount.parse().expect("Invalid ext_amount");
+            let abs_ext_amt = ext_amt.unsigned_abs();
+
+            if ext_amt.is_positive() {
+                return Err(Error::InvalidExecutionEntry);
+            } else {
+                if abs_ext_amt < self.min_withdraw_amt {
+                    return Err(Error::InvalidWithdrawAmount);
+                };
+
+                // TODO: execute token wrapper unwrap function
+            }
+
+            let fee_exists = ext_data_fee != 0;
+
+            if fee_exists {
+                if self
+                    .env()
+                    .transfer(ext_data.relayer.clone(), ext_data_fee)
```

Comment:
This should not be a native token transfer. Recall that the submitter of this transaction is a _relayer_. That relayer should be receiving the wrapped token that is being withdrawn and unshielded from the shielded pool.

Reply:
Alright. I will revert on this. This vanchor contract is not done yet until token-wrapper contract is ready.